cmd/operator/main.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	golog "log"
+	"time"
+
+	"github.com/golang/glog"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+
+	"github.com/openshift/hive/pkg/apis"
+	"github.com/openshift/hive/pkg/operator"
+
+	oappsv1 "github.com/openshift/api/apps/v1"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	apiregistrationv1 "k8s.io/kube-aggregator/pkg/apis/apiregistration/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
+
+	_ "github.com/openshift/generic-admission-server/pkg/cmd"
+)
+
+const (
+	defaultLogLevel = "info"
+)
+
+type controllerManagerOptions struct {
+	LogLevel string
+}
+
+func newRootCommand() *cobra.Command {
+	opts := &controllerManagerOptions{}
+	cmd := &cobra.Command{
+		Use:   "hive-operator",
+		Short: "OpenShift Hive Operator",
+		Run: func(cmd *cobra.Command, args []string) {
+			// Set log level
+			level, err := log.ParseLevel(opts.LogLevel)
+			if err != nil {
+				log.WithError(err).Fatal("Cannot parse log level")
+			}
+			log.SetLevel(level)
+			log.Debug("debug logging enabled")
+
+			// Get a config to talk to the apiserver
+			cfg, err := config.GetConfig()
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			// Create a new Cmd to provide shared dependencies and start components
+			mgr, err := manager.New(cfg, manager.Options{})
```

Comment:
It's a constant now in the operator code, hopefully that covers this comment but it not me know.

Reply:
Passing a namespace in Options scopes watches/caching to only that namespace, which makes it possible to watch for things like configmaps, secrets, etc. without requiring huge amounts of memory in your controller pod. Because you're still watching things like Deployments and Daemonsets, it'd be good to add the Namespace here so you don't cache all Deployments across the cluster.

Reply:
Ah now I understand. Added!